xbmc/video/VideoDatabase.cpp

```diff
@@ -3034,28 +3042,88 @@ bool CVideoDatabase::SetFileForEpisode(const std::string& fileAndPath, int idEpi
+bool CVideoDatabase::SetFileForMovie(const std::string& fileAndPath, int oldIdFile)
 {
+  const int idFile = AddFile(fileAndPath);
+  if (idFile < 0)
+    return false;
+
   try
   {
     assert(m_pDB->in_transaction());
+    std::string sql = PrepareSQL("UPDATE movie SET idFile=%i WHERE idFile=%i", idFile, oldIdFile);
+    m_pDS->exec(sql);
 
-    std::string sql = PrepareSQL("UPDATE movie SET c22='%s', idFile=%i WHERE idMovie=%i",
-                                 fileAndPath.c_str(), idFile, idMovie);
+    sql = PrepareSQL("UPDATE videoversion SET idFile=%i WHERE idFile=%i AND media_type='movie'",
+                     idFile, oldIdFile);
     m_pDS->exec(sql);
-    sql = PrepareSQL("UPDATE videoversion SET idFile=%i WHERE idMedia=%i AND media_type='movie'",
-                     idFile, idMovie);
+
+    sql = PrepareSQL("UPDATE art SET media_id=%i WHERE media_id=%i AND media_type='videoversion'",
+                     idFile, oldIdFile);
     m_pDS->exec(sql);
 
+    if (GetSingleValueInt(PrepareSQL("SELECT idFile FROM streamdetails WHERE idFile = %i", idFile),
+                          m_pDS) == 0)
+    {
+      sql = PrepareSQL("UPDATE streamdetails SET idFile=%i WHERE idFile=%i", idFile, oldIdFile);
+      m_pDS->exec(sql);
+    }
+    DeleteFile(oldIdFile);
+
     return true;
   }
   catch (...)
   {
-    CLog::Log(LOGERROR, "{} ({}) failed", __FUNCTION__, idMovie);
+    CLog::Log(LOGERROR, "{} ({}) failed", __FUNCTION__, idFile);
   }
   return false;
 }
 
+void CVideoDatabase::DeleteFile(int idFile)
+{
+  // First check not multiple references to file (eg. episodes)
+  std::string sql{PrepareSQL("SELECT idFile FROM movie WHERE idFile = %i UNION SELECT idFile FROM "
+                             "episode WHERE idFile = %i",
+                             idFile, idFile)};
+  m_pDS->query(sql);
+  if (m_pDS->eof())
+  {
+    // Get idPath
+    sql = PrepareSQL("SELECT path.idPath, path.strPath FROM path JOIN files ON path.idPath = "
+                     "files.idPath WHERE idFile = %i",
+                     idFile);
+    m_pDS->query(sql);
+    int path{-1};
+    std::string strPath;
+    if (!m_pDS->eof())
+    {
+      path = m_pDS->fv("idPath").get_asInt();
+      strPath = m_pDS->fv("strPath").get_asString();
+    }
+
+    // Remove file and any associated bookmarks and streamdetails
+    sql = PrepareSQL("DELETE FROM files WHERE idFile = %i", idFile);
+    m_pDS->exec(sql);
+    sql = PrepareSQL("DELETE FROM bookmark WHERE idFile = %i", idFile);
+    m_pDS->exec(sql);
+    DeleteStreamDetails(idFile);
+
+    // Delete path if orphan (and not base directory - needs to remain to prevent re-adding on library update)
+    if (path >= 0 && URIUtils::IsBluray(strPath))
+    {
+      if (GetSingleValueInt(
+              PrepareSQL("SELECT idFile FROM files JOIN path ON files.idPath = path.idPath WHERE "
+                         "files.idPath = %i",
+                         path),
+              m_pDS) == 0)
```

Comment:
not runtime tested, but this may return multiple rows, especially for episodes. Is that on purpose? not sure GetSingleValueInt is meant for that,
Is there a reason to include the path table in the query?